### PR TITLE
feat(settings): Add integrity checker

### PR DESCRIPTION
**Pull request by kiblik**
In many cases, users are changing `dojo/settings/settings.dist.py` file (they expect it is the way how the application should be configured).

This PR adds a checker that does not allow to start app if the change is detected. The error message points the user to documentation.

It is allowed only in debug mode when changes are expected.
Unittest which helps you in debug mode was added as well.

### CODE REVIEW

**Review by mtesauro — APPROVED** [review]
Approved

File: dojo/settings/settings.py [hunk 1]
```diff
@@ -7,3 +10,14 @@
...
     optional('local_settings.py')
 )
+
+if not DEBUG:
+    with (Path(__file__).parent / 'settings.dist.py').open('rb') as file:
+        real_hash = hashlib.sha256(file.read()).hexdigest()
+    with (Path(__file__).parent / '.settings.dist.py.sha256sum').open('rb') as file:
+        expected_hash = file.read().decode().strip()
+    if real_hash != expected_hash:
+        msg = "Change of 'settings.dist.py' file was detected. It is not allow to edit this file. " \
```
Comment (cneill) [inline dojo/settings/settings.py:20]:
```suggestion
        msg = "Change of 'settings.dist.py' file was detected. It is not allowed to edit this file. " \
```

File: unittests/test_utils.py [hunk 2]
```diff
@@ -240,3 +242,12 @@ def assertImportModelsCreated(test_case, tests=0, engagements=0, products=0, pro
...
               )
+
+
+class TestSettings(DojoTestCase):
+    def test_settings_integrity(self):
+        with Path('dojo/settings/settings.dist.py').open('rb') as file:
+            real_hash = hashlib.sha256(file.read()).hexdigest()
+        with Path('dojo/settings/.settings.dist.py.sha256sum').open('rb') as file:
+            expected_hash = file.read().decode().strip()
+        self.assertEqual(expected_hash, real_hash, "File settings.dist.py was changed but checksum has not been updated. Update sha256sum value in '.settings.dist.py.sha256sum'")
```
Comment (cneill) [inline unittests/test_utils.py:253]:
```suggestion
        self.assertEqual(expected_hash, real_hash, "File settings.dist.py was changed but checksum has not been updated. If this is part of a PR, update the sha256sum value in '.settings.dist.py.sha256sum'. If you are modifying this to configure your instance, revert your changes and use environment variables or 'local_settings.py'")
```

**Review by cneill — CHANGES_REQUESTED** [review]
Good idea. Just a few small suggestions. Given that people probably won't know about this if they're editing `settings.dist.py` as part of a PR, and we don't call this out explicitly for people who might be editing this on their instance, could you add a comment about this at the top of the file?

* Explain that this file shouldn't be edited for modifying a running instance (you can use the same message you used in the `ValueError`)
* Inform developers that they will need to update the hash in `.settings.dist.py.sha256sum` if they are modifying this file as part of a PR and explain how to do so (so they don't have to wait for a test failure to know about this)

**Comment by kiblik** [comment]
> Good idea. Just a few small suggestions. Given that people probably won't know about this if they're editing `settings.dist.py` as part of a PR, and we don't call this out explicitly for people who might be editing this on their instance, could you add a comment about this at the top of that file?
> 
> * Explain that this file shouldn't be edited for modifying a running instance (you can use the same message you used in the `ValueError`)
> * Inform developers that they will need to update the hash in `.settings.dist.py.sha256sum` if they are modifying this file as part of a PR and explain how to do so (so they don't have to wait for a test failure to know about this)

Thank you for your feedback. I agree.
I'm just not sure I understand where you would like to place these notes. Both on the top of the `settings.dist.py`? I'm not sure it is the best place because it increases the chances that also for non-developers will edit `settings.dist.py` and update the checksum.

**Review by cneill — APPROVED** [review]
The 2 comment blocks you added to `settings.dist.py` look good to me. Should resolve any confusion about who should and should not edit that file. Any thoughts @Maffooch @mtesauro ?

**Comment by mtesauro** [comment]
> The 2 comment blocks you added to `settings.dist.py` look good to me. Should resolve any confusion about who should and should not edit that file. Any thoughts @Maffooch @mtesauro ?

@cneill I'm good with the updates that @kiblik made in settings.dist.py :+1:

**Comment by cneill** [comment]
Sorry to ask for changes after approving, but since this will immediately break anyone who has modified their `settings.dist.py` file, would you mind adding a "breaking change" notice to the 2.35.0 release notes?

**Comment by kiblik** [comment]
> Sorry to ask for changes after approving, but since this will immediately break anyone who has modified their `settings.dist.py` file, would you mind adding a "breaking change" notice to the 2.35.0 release notes?

No problem. @cneill,  can you check the wording?

File: docs/content/en/getting_started/upgrading/2.35.md [hunk 3]
```diff
@@ -2,6 +2,11 @@
 title: 'Upgrading to DefectDojo Version 2.35.x'
 toc_hide: true
 weight: -20240506
-description: No special instructions.
+description: Integrity checker announced
 ---
-There are no special instructions for upgrading to 2.35.x. Check the [Release Notes](https://github.com/DefectDojo/django-DefectDojo/releases/tag/2.35.0) for the contents of the release.
+
+From 2.35.0, DefectDojo is performing integrity check of `settings.dist.py` file. If the user changed this file (in the past or even now) DefectDojo will avoid starting of instance.
```
Comment (cneill) [inline docs/content/en/getting_started/upgrading/2.35.md:8]:
```suggestion
From 2.35.0, DefectDojo will perform an integrity check of the `settings.dist.py` file to ensure it has not been modified. If the user changed this file (in the past or even now) the DefectDojo instance will not start until those changes have been reverted.
```

**Review by cneill — COMMENTED** [review]
Just a few small suggestions. Thanks for adding this

**Review by grendel513 — APPROVED** [review]

**Comment by quirinziessler** [comment]
@kiblik how extensive have you tested this? I am not able to build even I enabled the Debug mode.

`❯ docker/setEnv.sh debug
Now using 'debug' configuration.
❯ ./dc-build.sh
Checking docker compose version
Supported docker compose version
Building docker compose
[+] Building 0.7s (35/39)                                                                                          docker:colima
 => [uwsgi internal] load build definition from Dockerfile.django-debian                                                    0.0s
 => => transferring dockerfile: 4.81kB                                                                                      0.0s
 => [uwsgi internal] load .dockerignore                                                                                     0.0s
 => => transferring context: 98B                                                                                            0.0s
 => [nginx internal] load metadata for docker.io/library/python:3.11.4-slim-bullseye@sha256:40319d0a897896e746edf877783ef3  0.0s
 => [nginx base 1/1] FROM docker.io/library/python:3.11.4-slim-bullseye@sha256:40319d0a897896e746edf877783ef39685d44e90e1e  0.0s
 => [uwsgi internal] load build context                                                                                     0.2s
 => => transferring context: 2.19MB                                                                                         0.2s
 => CACHED [nginx build 1/4] WORKDIR /app                                                                                   0.0s
 => CACHED [uwsgi django  2/12] RUN   apt-get -y update &&   mkdir -p /usr/share/man/man1 /usr/share/man/man7 &&   apt-get  0.0s
 => CACHED [nginx build 2/4] RUN   apt-get -y update &&   apt-get -y install --no-install-recommends     gcc     build-ess  0.0s
 => CACHED [uwsgi build 3/4] COPY requirements.txt ./                                                                       0.0s
 => CACHED [uwsgi build 4/4] RUN CPUCOUNT=1 pip3 wheel --wheel-dir=/tmp/wheels -r ./requirements.txt                        0.0s
 => CACHED [uwsgi django  3/12] COPY --from=build /tmp/wheels /tmp/wheels                                                   0.0s
 => CACHED [uwsgi django  4/12] COPY requirements.txt ./                                                                    0.0s
 => CACHED [uwsgi django  5/12] RUN export PYCURL_SSL_LIBRARY=openssl &&     pip3 install  --no-cache-dir  --no-index   --  0.0s
 => CACHED [uwsgi django  6/12] COPY   docker/entrypoint-celery-beat.sh   docker/entrypoint-celery-worker.sh   docker/entr  0.0s
 => CACHED [uwsgi django  7/12] COPY wsgi.py manage.py docker/unit-tests.sh ./                                              0.0s
 => CACHED [uwsgi django  8/12] COPY dojo/ ./dojo/                                                                          0.0s
 => CACHED [uwsgi django  9/12] COPY docker/extra_fixtures/* /app/dojo/fixtures/                                            0.0s
 => CACHED [uwsgi django 10/12] COPY tests/ ./tests/                                                                        0.0s
 => CACHED [uwsgi django 11/12] RUN   rm -f /readme.txt &&   rm -f dojo/fixtures/readme.txt &&   mkdir -p dojo/migrations   0.0s
 => CACHED [uwsgi django 12/12] RUN     addgroup --gid 1337 defectdojo &&     adduser --system --no-create-home --disabled  0.0s
 => [uwsgi] exporting to image                                                                                              0.0s
 => => exporting layers                                                                                                     0.0s
 => => writing image sha256:970d0ef2eab758b9b7cb5801c1e21805628d2da39f8242d7c2d049e369106c5d                                0.0s
 => => naming to docker.io/defectdojo/defectdojo-django:latest                                                              0.0s
 => [nginx internal] load build definition from Dockerfile.nginx-debian                                                     0.0s
 => => transferring dockerfile: 3.46kB                                                                                      0.0s
 => [nginx internal] load .dockerignore                                                                                     0.0s
 => => transferring context: 98B                                                                                            0.0s
 => [nginx internal] load metadata for docker.io/library/nginx:1.26.0-alpine@sha256:ef587d1eb99e991291c582bfb74f27db27f7ca  0.0s
 => [nginx internal] load build context                                                                                     0.1s
 => => transferring context: 263.65kB                                                                                       0.1s
 => CACHED [nginx stage-3 1/5] FROM docker.io/library/nginx:1.26.0-alpine@sha256:ef587d1eb99e991291c582bfb74f27db27f7ca2c0  0.0s
 => CACHED [nginx build 3/4] COPY requirements.txt ./                                                                       0.0s
 => CACHED [nginx build 4/4] RUN CPUCOUNT=1 pip3 wheel --wheel-dir=/tmp/wheels -r ./requirements.txt                        0.0s
 => CACHED [nginx collectstatic 1/7] RUN   apt-get -y update &&   apt-get -y install --no-install-recommends apt-transport  0.0s
 => CACHED [nginx collectstatic 2/7] RUN pip3 install  --no-cache-dir  --no-index   --find-links=/tmp/wheels  -r ./require  0.0s
 => CACHED [nginx collectstatic 3/7] COPY components/ ./components/                                                         0.0s
 => CACHED [nginx collectstatic 4/7] RUN   cd components &&   yarn                                                          0.0s
 => CACHED [nginx collectstatic 5/7] COPY manage.py ./                                                                      0.0s
 => CACHED [nginx collectstatic 6/7] COPY dojo/ ./dojo/                                                                     0.0s
 => ERROR [nginx collectstatic 7/7] RUN env DD_SECRET_KEY='.' python3 manage.py collectstatic --noinput && true             0.2s
------
 > [nginx collectstatic 7/7] RUN env DD_SECRET_KEY='.' python3 manage.py collectstatic --noinput && true:
0.210 Traceback (most recent call last):
0.210   File "/app/manage.py", line 10, in <module>
0.210     execute_from_command_line(sys.argv)
0.210   File "/usr/local/lib/python3.11/site-packages/django/core/management/__init__.py", line 446, in execute_from_command_line
0.210     utility.execute()
0.210   File "/usr/local/lib/python3.11/site-packages/django/core/management/__init__.py", line 386, in execute
0.210     settings.INSTALLED_APPS
0.210   File "/usr/local/lib/python3.11/site-packages/django/conf/__init__.py", line 92, in __getattr__
0.210     self._setup(name)
0.210   File "/usr/local/lib/python3.11/site-packages/django/conf/__init__.py", line 79, in _setup
0.210     self._wrapped = Settings(settings_module)
0.210                     ^^^^^^^^^^^^^^^^^^^^^^^^^
0.210   File "/usr/local/lib/python3.11/site-packages/django/conf/__init__.py", line 190, in __init__
0.210     mod = importlib.import_module(self.SETTINGS_MODULE)
0.210           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
0.210   File "/usr/local/lib/python3.11/importlib/__init__.py", line 126, in import_module
0.210     return _bootstrap._gcd_import(name[level:], package, level)
0.210            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
0.210   File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
0.210   File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
0.210   File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
0.210   File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
0.210   File "<frozen importlib._bootstrap_external>", line 940, in exec_module
0.210   File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
0.210   File "/app/dojo/settings/settings.py", line 23, in <module>
0.210     raise ValueError(msg)
0.210 ValueError: Change of 'settings.dist.py' file was detected. It is not allowed to edit this file. Any customization of variables need to be done via environmental variables or in 'local_settings.py'. For more information check https://documentation.defectdojo.com/getting_started/configuration/ 
------
failed to solve: process "/bin/sh -c env DD_SECRET_KEY='.' python3 manage.py collectstatic --noinput && true" did not complete successfully: exit code: 1`

**Comment by kiblik** [comment]
> @kiblik how extensive have you tested this? I am not able to build even I enabled the Debug mode.
> 
I have tested the local change of `settings.dist.py` (with a prebuilt image) and I tested the build in GitHub workflow. You are right that I forgot about the local `./dc-build.sh` option with `manage.py collectstatic`.

I fixed this issue now: #10241

**Pull request merged**